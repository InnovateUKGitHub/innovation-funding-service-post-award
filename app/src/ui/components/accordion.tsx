import * as React from "react";

interface Props {
  title: string;
  content: JSX.Element;
}

interface State {
  accordionOpen: boolean;
}

export class Accordion extends React.Component<Props, State> {

  constructor(props: Props) {
    super(props);
    this.state = {
      accordionOpen: false
    };
  }

  render() {
    const styles = {
      buttonStyle: {
        background: "none",
        border: "none"
      }
    };

    return (
      <div className="govuk-accordion__section">
        <div className="govuk-accordion__section-header" onClick={() => this.setState({accordionOpen: !this.state.accordionOpen})}>
          <button style={styles.buttonStyle} className="govuk-heading-m govuk-!-margin-bottom-3 govuk-!-margin-top-3">{this.props.title}</button>
          <img className="govuk-!-padding-right-2" src={this.state.accordionOpen ? "/assets/images/icon-minus.png" : "/assets/images/icon-plus.png"}/>
        </div>
        {this.state.accordionOpen && <div className="govuk-accordion__section-panel">
          {this.props.content}
        </div>}
      </div>
    );
  }
}
