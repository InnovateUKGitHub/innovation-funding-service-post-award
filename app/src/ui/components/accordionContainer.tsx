import * as React from "react";

interface Props {
  children: JSX.Element | JSX.Element[];
}

export class AccordionContainer extends React.Component<Props> {

  render() {
    return (
      <div className="govuk-accordion">
        {this.props.children}
      </div>
    );
  }
}
