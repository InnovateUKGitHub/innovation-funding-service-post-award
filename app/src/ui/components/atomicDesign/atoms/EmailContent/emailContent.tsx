import type { ContentSelector } from "@copy/type";
import { useContent } from "@ui/hooks/content.hook";
import { EmailProps, Email } from "../Email/Email";

export interface EmailContentProps extends Omit<EmailProps, "value" | "children"> {
  value: ContentSelector;
  qa?: string;
}

/**
 * Wrapper component for the `<Email />` component
 * that allows passing in content using the content
 * function as a value.
 *
 * @example
 * <EmailContent value={x => x.myPage.emailAddress} />
 */
export function EmailContent({ value, ...props }: EmailContentProps) {
  const { getContent } = useContent();
  const email = getContent(value);

  return <Email {...props}>{email || ""}</Email>;
}
