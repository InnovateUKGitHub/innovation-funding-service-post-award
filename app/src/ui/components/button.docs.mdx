import { Meta, Primary, Controls, Canvas, Story } from "@storybook/blocks";
import * as ButtonStories from "./button.stories";

<Meta of={ButtonStories} />

# Button

> Related links:
>
> - [GOV.UK Design System on Button](https://design-system.service.gov.uk/components/button/)
> - [Take part in the 'Button' discusson on GitHub and share your research](https://github.com/alphagov/govuk-design-system-backlog/issues/34)

<Primary />
<Controls />

## Scenarios

### Default Buttons

<Canvas>
  <Story of={ButtonStories.PrimaryButton} />
</Canvas>

### Start Buttons

Start Buttons have not been implemented by IFSPA.

### Secondary Buttons

<Canvas>
  <Story of={ButtonStories.SecondaryButton} />
</Canvas>

### Warning Buttons

<Canvas>
  <Story of={ButtonStories.WarningButton} />
</Canvas>

### Disabled Buttons

Disabled buttons have poor contrast and can confuse some users, so avoid them if possible.

<Canvas>
  <Story of={ButtonStories.DisabledButton} />
</Canvas>

### Link Buttons

Link buttons are not GOV.UK components;
They are used in IFSPA in place of a button where buttons would be "too crowded",
such as within lists of documents, where each row requires a delete button.

Do not use this component in place of a link.
Use this component when a button would suffice.

<Canvas>
  <Story of={ButtonStories.LinkButton} />
</Canvas>
