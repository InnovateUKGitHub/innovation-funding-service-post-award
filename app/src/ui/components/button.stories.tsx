import type { Meta, StoryObj } from "@storybook/react";

import { Button } from "./styledButton";

const meta: Meta<typeof Button> = {
  title: "GOV.UK Components/Button",
  component: Button,
  argTypes: {
    styling: {
      defaultValue: "Primary",
      description: "The apperance of the button",
    },
    disabled: {
      defaultValue: false,
      description: "Whether the button is disabled",
      type: "boolean",
    },
  },
};

export default meta;

type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  tags: ["govuk/component"],
  args: {
    styling: "Primary",
    children: "Return to Desktop",
  },
};

export const PrimaryButton: Story = {
  tags: ["govuk/component"],
  args: {
    styling: "Primary",
    children: "A primary button",
  },
};

export const DisabledButton: Story = {
  tags: ["govuk/component"],
  args: {
    styling: "Primary",
    disabled: true,
    children: "A disabled button",
  },
};

export const SecondaryButton: Story = {
  tags: ["govuk/component"],
  args: {
    styling: "Secondary",
    children: "A secondary button",
  },
};

export const WarningButton: Story = {
  tags: ["govuk/component"],
  args: {
    styling: "Warning",
    children: "A warning button",
  },
};

export const LinkButton: Story = {
  tags: ["ifspa/component"],
  args: {
    styling: "Link",
    children: "A link button",
  },
};
