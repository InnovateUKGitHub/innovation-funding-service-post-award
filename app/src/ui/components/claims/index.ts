export * from "./claimDetailsLink";
export * from "./claimLastModified";
export * from "./AwardRateOverridesMessage";
export * from "./claimPeriodDate";
export * from "./claimTable";
export * from "./claimReviewTable";
export * from "./forecastTable";
