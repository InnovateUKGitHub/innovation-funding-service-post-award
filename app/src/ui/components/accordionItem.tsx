import * as React from "react";
import cn from "classnames";
import { AccordionContext } from "./accordion";

interface Props {
  title: string;
  qa?: string;
}

interface State {
  accordionOpen: boolean;
  focused: boolean;
}

export class AccordionItem extends React.Component<Props, State> {

  static contextType = AccordionContext;
  context!: React.ContextType<typeof AccordionContext>;

  constructor(props: Props) {
    super(props);
    this.state = {
      accordionOpen: true, // on server needs to be open
      focused: false
    };
  }

  componentDidMount() {
    this.context.subscribe();
    this.setState({ accordionOpen: false }); // once mounted on client can be initally shut
  }

  onClick() {
    const newAccordionState = !this.state.accordionOpen;
    this.context.toggle(newAccordionState);
    this.setState({ accordionOpen: newAccordionState });
  }

  onBlur() {
    this.setState({ focused: false });
  }

  onFocus() {
    this.setState({ focused: true });
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (this.context.allOpen && !prevState.accordionOpen) {
      this.setState({
        accordionOpen: true
      });
    }
    if (this.context.allClosed && prevState.accordionOpen) {
      this.setState({
        accordionOpen: false
      });
    }
  }

  renderTitle() {
    if (!this.context.jsEnabled) {
      return <span className="govuk-accordion__section-button">{this.props.title}</span>;
    }
    return (
      <React.Fragment>
        <button type="button" className="govuk-accordion__section-button" aria-expanded={this.state.accordionOpen} data-qa={`${this.props.qa}-button`}>{this.props.title}</button>
        <span className="govuk-accordion__icon" aria-hidden={!this.state.accordionOpen}/>
      </React.Fragment>
    );
  }

  render() {
    const topLevelClasses = cn({
      "govuk-accordion__section": true,
      "govuk-accordion__section--expanded": this.state.accordionOpen
    });
    const headerClasses = cn({
      "govuk-accordion__section-header": true,
      "govuk-accordion__section-header--focused": this.state.focused
    });
    return (
      <div className={topLevelClasses} data-qa={this.props.qa}>
        <div onBlur={() => this.onBlur()} onFocus={() => this.onFocus()} onClick={() => this.onClick()} className={headerClasses} >
          <p className="govuk-accordion__section-heading">
            {this.renderTitle()}
          </p>
        </div>
        <div className="govuk-accordion__section-content">
          {this.props.children}
        </div>
      </div>
    );
  }
}
