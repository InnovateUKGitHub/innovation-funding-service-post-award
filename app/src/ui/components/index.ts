import * as Renderers from "./renderers";
import * as Projects from "./projects";
import * as Claims from "./claims";
import * as Inputs from "./inputs";
import * as MonitoringReports from "./monitoringReports";

export * from "./accordion";
export * from "./accordionItem";
export * from "./details";
export * from "./documentList";
export * from "./documentSingle";
export * from "./errorSummary";
export * from "./form";
export * from "./layout";
export * from "./links";
export * from "./linksList";
export * from "./loading";
export * from "./logs";
export * from "./navigationArrows";
export * from "./partnersAndFinanceContacts";
export * from "./periodTitle";
export * from "./projectMember";
export * from "./styledButton";
export * from "./table";
export * from "./validationError";
export * from "./validationMessage";
export * from "./validationSummary";

export { Renderers, Projects, Claims, Inputs, MonitoringReports };
