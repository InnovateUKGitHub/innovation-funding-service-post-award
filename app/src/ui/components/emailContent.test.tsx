import TestBed, { TestBedContent } from "@shared/TestBed";
import { render } from "@testing-library/react";

import { EmailContent, EmailContentProps } from "@ui/components/emailContent";

describe("EmailContent", () => {
  const content = {
    projectDetails: {
      changeEmail: {
        content: "testbed@iuk.ukri.org",
      },
    },
  } as TestBedContent;
  const setup = (props?: Partial<EmailContentProps>) =>
    render(
      <TestBed content={content}>
        <EmailContent value={x => x.projectDetails.changeEmail} {...props} />
      </TestBed>,
    );

  describe("@renders", () => {
    test("should return as default", () => {
      const { container } = setup({ qa: "email-qa" });
      expect(container).toMatchSnapshot();
    });

    describe("should render href", () => {
      test("with default value", () => {
        const { getByRole } = setup();
        const expectedHref = `mailto:${content.projectDetails?.changeEmail.content}`;
        const anchorEl = getByRole("link");
        expect(anchorEl).toHaveAttribute("href", expectedHref);
      });

      test("with provided href", () => {
        const { getByRole } = setup({ href: "predefinedhref@test.com" });
        const anchorEl = getByRole("link");
        expect(anchorEl.getAttribute("href")).toEqual("mailto:predefinedhref@test.com");
      });
    });

    test("with classname", () => {
      const { getByRole } = setup({ className: "customClass" });
      const anchorEl = getByRole("link");
      expect(anchorEl).toHaveClass("customClass");
    });

    test("with qa", () => {
      const { getByRole } = setup({ qa: "email-qa" });
      const anchorEl = getByRole("link");
      expect(anchorEl).toHaveAttribute("data-qa");
      expect(anchorEl.getAttribute("data-qa")).toBe("email-qa");
    });

    test("when empty should render null", () => {
      const { queryByRole } = setup({ value: undefined });
      expect(queryByRole("link")).toBe(null);
    });
  });
});
