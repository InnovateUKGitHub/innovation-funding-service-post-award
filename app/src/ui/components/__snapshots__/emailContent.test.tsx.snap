// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailContent @renders should return as default 1`] = `
<div>
  <a
    class="govuk-link govuk-!-font-size-19"
    data-qa="email-qa"
    href="mailto:testbed@iuk.ukri.org"
  >
    testbed@iuk.ukri.org
  </a>
</div>
`;
